File: src/test/java/com/flexpoker/game/command/aggregate/tablebalancer/TableBalancerNothingToDoTest.java
package com.flexpoker.game.command.aggregate.tablebalancer;

import static com.flexpoker.game.command.aggregate.tablebalancer.TableBalancerTestUtils.createDefaultChipMapForSubjectTable;
import static com.flexpoker.game.command.aggregate.tablebalancer.TableBalancerTestUtils.createTableToPlayersMap;
import static org.junit.jupiter.api.Assertions.assertFalse;

import java.util.Collections;
import java.util.UUID;

import org.junit.jupiter.api.Test;

import com.flexpoker.game.command.aggregate.TableBalancer;

public class TableBalancerNothingToDoTest {

    @Test
    void testSingleTableTwoPlayers() {
        var subjectTableId = UUID.randomUUID();
        var tableToPlayersMap = createTableToPlayersMap(subjectTableId, 2);

        var tableBalancer = new TableBalancer(UUID.randomUUID(), 2);
        var event = tableBalancer.createSingleBalancingEvent(subjectTableId, Collections.emptySet(),
                tableToPlayersMap, createDefaultChipMapForSubjectTable(subjectTableId, tableToPlayersMap));
        assertFalse(event.isPresent());
    }

    @Test
    void testTwoTablesTwoPlayersEach() {
        var subjectTableId = UUID.randomUUID();
        var tableToPlayersMap = createTableToPlayersMap(subjectTableId, 2, 2);

        var tableBalancer = new TableBalancer(UUID.randomUUID(), 2);
        var event = tableBalancer.createSingleBalancingEvent(subjectTableId, Collections.emptySet(),
                tableToPlayersMap, createDefaultChipMapForSubjectTable(subjectTableId, tableToPlayersMap));
        assertFalse(event.isPresent());
    }

    @Test
    void testTwoTablesOneAlreadyPausedWaitingForMerge() {
        var subjectTableId = UUID.randomUUID();
        var tableToPlayersMap = createTableToPlayersMap(subjectTableId, 2, 1);
        var otherTableId = tableToPlayersMap.keySet().stream()
                .filter(x -> !x.equals(subjectTableId)).findFirst().get();

        var tableBalancer = new TableBalancer(UUID.randomUUID(), 2);
        var event = tableBalancer.createSingleBalancingEvent(subjectTableId, Collections.singleton(otherTableId),
                tableToPlayersMap, createDefaultChipMapForSubjectTable(subjectTableId, tableToPlayersMap));
        assertFalse(event.isPresent());
    }

    @Test
    void testTwoTablesWithinOneOfEachOtherAndUnderMergeThreshold() {
        var subjectTableId = UUID.randomUUID();
        var tableToPlayersMap = createTableToPlayersMap(subjectTableId, 6, 7);

        var tableBalancer = new TableBalancer(UUID.randomUUID(), 9);
        var event = tableBalancer.createSingleBalancingEvent(subjectTableId, Collections.emptySet(),
                tableToPlayersMap, createDefaultChipMapForSubjectTable(subjectTableId, tableToPlayersMap));
        assertFalse(event.isPresent());
    }

    @Test
    void testThreeTablesPerfectlyInBalanceAndUnderMergeThreshold() {
        var subjectTableId = UUID.randomUUID();
        var tableToPlayersMap = createTableToPlayersMap(subjectTableId, 7, 7, 7);

        var tableBalancer = new TableBalancer(UUID.randomUUID(), 9);
        var event = tableBalancer.createSingleBalancingEvent(subjectTableId, Collections.emptySet(),
                tableToPlayersMap, createDefaultChipMapForSubjectTable(subjectTableId, tableToPlayersMap));
        assertFalse(event.isPresent());
    }

    @Test
    void testThreeTablesAllAtMax() {
        var subjectTableId = UUID.randomUUID();
        var tableToPlayersMap = createTableToPlayersMap(subjectTableId, 6, 6, 6);

        var tableBalancer = new TableBalancer(UUID.randomUUID(), 6);
        var event = tableBalancer.createSingleBalancingEvent(subjectTableId, Collections.emptySet(),
                tableToPlayersMap, createDefaultChipMapForSubjectTable(subjectTableId, tableToPlayersMap));
        assertFalse(event.isPresent());
    }

    @Test
    void testThreeTablesTwoImbalancedByTwoCantMergeSubjectIsMediumTableSmallestPaused() {
        var subjectTableId = UUID.randomUUID();
        var tableToPlayersMap = createTableToPlayersMap(subjectTableId, 8, 7, 9);
        var smallestOtherTableId = tableToPlayersMap.entrySet().stream()
                .filter(x -> x.getValue().size() == 7).findFirst().get()
                .getKey();

        var tableBalancer = new TableBalancer(UUID.randomUUID(), 9);
        var event = tableBalancer.createSingleBalancingEvent(subjectTableId,
                Collections.singleton(smallestOtherTableId), tableToPlayersMap,
                createDefaultChipMapForSubjectTable(subjectTableId, tableToPlayersMap));
        assertFalse(event.isPresent());
    }

    @Test
    void testThreeTablesAllImbalancedCantMergeSubjectIsMediumTableSmallestPaused() {
        var subjectTableId = UUID.randomUUID();
        var tableToPlayersMap = createTableToPlayersMap(subjectTableId, 7, 6, 9);
        var smallestOtherTableId = tableToPlayersMap.entrySet().stream()
                .filter(x -> x.getValue().size() == 6).findFirst().get()
                .getKey();

        var tableBalancer = new TableBalancer(UUID.randomUUID(), 9);
        var event = tableBalancer.createSingleBalancingEvent(subjectTableId,
                Collections.singleton(smallestOtherTableId), tableToPlayersMap,
                createDefaultChipMapForSubjectTable(subjectTableId, tableToPlayersMap));
        assertFalse(event.isPresent());
    }

}

File: src/test/kotlin/com/flexpoker/game/command/aggregate/tablebalancer/TableBalancerNothingToDoTest.kt
package com.flexpoker.game.command.aggregate.tablebalancer

import com.flexpoker.game.command.aggregate.createSingleBalancingEvent
import com.flexpoker.test.util.UnitTestClass
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Test
import java.util.UUID

@UnitTestClass
class TableBalancerNothingToDoTest {

    @Test
    fun testSingleTableTwoPlayers() {
        val subjectTableId = UUID.randomUUID()
        val tableToPlayersMap = TableBalancerTestUtils.createTableToPlayersMap(subjectTableId, 2)
        val event = createSingleBalancingEvent(UUID.randomUUID(), 2, subjectTableId, emptySet(), tableToPlayersMap,
            TableBalancerTestUtils.createDefaultChipMapForSubjectTable(subjectTableId, tableToPlayersMap))
        assertFalse(event.isPresent)
    }

    @Test
    fun testTwoTablesTwoPlayersEach() {
        val subjectTableId = UUID.randomUUID()
        val tableToPlayersMap = TableBalancerTestUtils.createTableToPlayersMap(subjectTableId, 2, 2)
        val event = createSingleBalancingEvent(UUID.randomUUID(), 2, subjectTableId, emptySet(),
            tableToPlayersMap, TableBalancerTestUtils.createDefaultChipMapForSubjectTable(subjectTableId, tableToPlayersMap))
        assertFalse(event.isPresent)
    }

    @Test
    fun testTwoTablesOneAlreadyPausedWaitingForMerge() {
        val subjectTableId = UUID.randomUUID()
        val tableToPlayersMap = TableBalancerTestUtils.createTableToPlayersMap(subjectTableId, 2, 1)
        val otherTableId = tableToPlayersMap.keys.first { it != subjectTableId }
        val event = createSingleBalancingEvent(UUID.randomUUID(), 2, subjectTableId, setOf(otherTableId),
            tableToPlayersMap, TableBalancerTestUtils.createDefaultChipMapForSubjectTable(subjectTableId, tableToPlayersMap))
        assertFalse(event.isPresent)
    }

    @Test
    fun testTwoTablesWithinOneOfEachOtherAndUnderMergeThreshold() {
        val subjectTableId = UUID.randomUUID()
        val tableToPlayersMap = TableBalancerTestUtils.createTableToPlayersMap(subjectTableId, 6, 7)
        val event = createSingleBalancingEvent(UUID.randomUUID(), 9, subjectTableId, emptySet(),
            tableToPlayersMap, TableBalancerTestUtils.createDefaultChipMapForSubjectTable(subjectTableId, tableToPlayersMap))
        assertFalse(event.isPresent)
    }

    @Test
    fun testThreeTablesPerfectlyInBalanceAndUnderMergeThreshold() {
        val subjectTableId = UUID.randomUUID()
        val tableToPlayersMap = TableBalancerTestUtils.createTableToPlayersMap(subjectTableId, 7, 7, 7)
        val event = createSingleBalancingEvent(UUID.randomUUID(), 9, subjectTableId, emptySet(),
            tableToPlayersMap, TableBalancerTestUtils.createDefaultChipMapForSubjectTable(subjectTableId, tableToPlayersMap))
        assertFalse(event.isPresent)
    }

    @Test
    fun testThreeTablesAllAtMax() {
        val subjectTableId = UUID.randomUUID()
        val tableToPlayersMap = TableBalancerTestUtils.createTableToPlayersMap(subjectTableId, 6, 6, 6)
        val event = createSingleBalancingEvent(UUID.randomUUID(), 6, subjectTableId, emptySet(),
            tableToPlayersMap, TableBalancerTestUtils.createDefaultChipMapForSubjectTable(subjectTableId, tableToPlayersMap))
        assertFalse(event.isPresent)
    }

    @Test
    fun testThreeTablesTwoImbalancedByTwoCantMergeSubjectIsMediumTableSmallestPaused() {
        val subjectTableId = UUID.randomUUID()
        val tableToPlayersMap = TableBalancerTestUtils.createTableToPlayersMap(subjectTableId, 8, 7, 9)
        val smallestOtherTableId = tableToPlayersMap.entries.first { it.value.size == 7 }.key
        val event = createSingleBalancingEvent(UUID.randomUUID(), 9, subjectTableId, setOf(smallestOtherTableId),
            tableToPlayersMap, TableBalancerTestUtils.createDefaultChipMapForSubjectTable(subjectTableId, tableToPlayersMap))
        assertFalse(event.isPresent)
    }

    @Test
    fun testThreeTablesAllImbalancedCantMergeSubjectIsMediumTableSmallestPaused() {
        val subjectTableId = UUID.randomUUID()
        val tableToPlayersMap = TableBalancerTestUtils.createTableToPlayersMap(subjectTableId, 7, 6, 9)
        val smallestOtherTableId = tableToPlayersMap.entries.first { it.value.size == 6 }.key
        val event = createSingleBalancingEvent(UUID.randomUUID(), 9, subjectTableId, setOf(smallestOtherTableId),
            tableToPlayersMap, TableBalancerTestUtils.createDefaultChipMapForSubjectTable(subjectTableId, tableToPlayersMap))
        assertFalse(event.isPresent)
    }

}